chia/util/action_scope.py

```diff
@@ -123,32 +53,37 @@ def config(self) -> _T_Config:
     @contextlib.asynccontextmanager
     async def new_scope(
         cls,
-        side_effects_format: type[_T_SideEffects],
+        lock: Callable[..., contextlib.AbstractAsyncContextManager[Any]],
+        initial_side_effects: _T_SideEffects,
         # I want a default here in case a use case doesn't want to take advantage of the config but no default seems to
         # satisfy the type hint _T_Config so we'll just ignore this.
         config: _T_Config = object(),  # type: ignore[assignment]
-        resource_manager_backend: type[ResourceManager] = SQLiteResourceManager,
     ) -> AsyncIterator[ActionScope[_T_SideEffects, _T_Config]]:
-        async with resource_manager_backend.managed(side_effects_format()) as resource_manager:
-            self = cls(_resource_manager=resource_manager, _side_effects_format=side_effects_format, _config=config)
+        self = cls(
+            _lock=lock, _active_interface=StateInterface(initial_side_effects, _callbacks_allowed=True), _config=config
+        )
 
-            yield self
+        yield self
 
-            async with self.use(_callbacks_allowed=False) as interface:
-                if self._callback is not None:
-                    await self._callback(interface)
-                self._final_side_effects = interface.side_effects
+        async with self.use(_callbacks_allowed=False) as interface:
+            if interface.callback is not None:
+                await interface.callback(interface)
+            self._final_side_effects = interface.side_effects
 
     @contextlib.asynccontextmanager
     async def use(self, _callbacks_allowed: bool = True) -> AsyncIterator[StateInterface[_T_SideEffects]]:
-        async with self._resource_manager.use():
-            side_effects = await self._resource_manager.get_resource(self._side_effects_format)
-            interface = StateInterface(side_effects, _callbacks_allowed, self._callback)
-
-            yield interface
-
-            await self._resource_manager.save_resource(interface.side_effects)
-            self._callback = interface.callback
+        new_interface = copy.deepcopy(self._active_interface)
+        previous_interface = self._active_interface
+
+        try:
+            self._active_interface = new_interface
+            self._active_interface._callbacks_allowed = _callbacks_allowed
+            yield self._active_interface
+            self._active_interface._callbacks_allowed = previous_interface._callbacks_allowed
+            for field, value in self._active_interface.__dict__.items():
+                previous_interface.__dict__[field] = value
+        finally:
+            self._active_interface = previous_interface
```

Comment:
Do we need three copies?  Previous, new, being-updated-from-previous?  Or, a maybe a we-are-damaged marker that blocks further usage in case the local actions here fail and the caller consumes the failue and keeps on going.